File: java/SwapNodesinPairs.java
/*
 * Filename: SwapNodesinPairs.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 13 Mar 2015 12:15:29 PM EDT
 */

/*
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
	public ListNode swapPairs(ListNode head) {
		ListNode dummy = new ListNode(0);
		dummy.next = head;

		ListNode prev = dummy;
		ListNode curr = head;
		ListNode temp;

		while (curr != null && curr.next != null) {
			prev.next = curr.next;
			prev = prev.next;

			temp = curr;
			curr.next = prev.next;
			curr = curr.next;

			prev.next = temp;
			prev = prev.next;
		}

		return dummy.next;
	}
}

File: cpp/swapNodesInPairs.cpp
/*
 * Filename: swapNodesInPairs.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Mon 11 Jan 2016 09:44:04 PM EST
 * Description: 
 */

#include <iostream>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
	public:
		ListNode* swapPairs(ListNode* head)
		{
			ListNode* res = new ListNode(0);
			ListNode* dummy = res;
			ListNode* p;
			ListNode* q;

			while (head && head->next)
			{
				p = head;
				q = head->next;
				head = head->next->next;

				res->next = q;
				q->next = p;
				p->next = NULL;
				res = res->next->next;
			}

			if (head)
			{
				res->next = head;
			}

			return dummy->next;
		}
};
